src/main/scala/internal/Graph.scala

```diff
@@ -0,0 +1,425 @@
+
+    val id = nextSinkId()
+    val newSinkDependencies = sinkDependencies.updated(id, dep)
+    (copy(sinkDependencies = newSinkDependencies), id)
+  }
+
+  /**
+   * Immutably add a new source to this graph.
+   *
+   * @return A pair containing the new graph and the id assigned to the new source
+   */
+  def addSource(): (Graph, SourceId) = {
+    val id = nextSourceId()
+    val newSources = sources + id
+    (copy(sources = newSources), id)
+  }
+
+  /**
+   * Immutably update the dependencies of a node in this graph.
+   *
+   * @param node The id of the node to be updated
+   * @param deps The new dependencies to assign to the node
+   * @return The new graph
+   */
+  def setDependencies(node: NodeId, deps: Seq[NodeOrSourceId]): Graph = {
+    require(dependencies.contains(node), "Node being updated must exist")
+    require ({
+      val nodesAndSources: Set[NodeOrSourceId] = nodes ++ sources
+      deps.forall(dep => nodesAndSources.contains(dep))
+    }, "Node must have dependencies on existing ids")
+
+    val newDependencies = dependencies.updated(node, deps)
+    copy(dependencies = newDependencies)
+  }
+
+  /**
+   * Immutably update the operator of a node in this graph.
+   *
+   * @param node The id of the node to be updated
+   * @param op The new operator to assign to the node
+   * @return The new graph
+   */
+  def setOperator(node: NodeId, op: Operator): Graph = {
+    require(dependencies.contains(node), "Node being updated must exist")
+
+    val newOperators = operators.updated(node, op)
+    copy(operators = newOperators)
+  }
+
+  /**
+   * Immutably update the dependency of a sink in this graph.
+   *
+   * @param sink The id of the sink to be updated
+   * @param dep The new dependency to assign to the sink
+   * @return The new graph
+   */
+  def setSinkDependency(sink: SinkId, dep: NodeOrSourceId): Graph = {
+    require(sinkDependencies.contains(sink), "Sink being updated must exist")
+
+    require ({
+      val nodesAndSources: Set[NodeOrSourceId] = nodes ++ sources
+      nodesAndSources.contains(dep)
+    }, "Sink must have dependencies on an existing id")
+
+    val newSinkDependencies = sinkDependencies.updated(sink, dep)
+    copy(sinkDependencies = newSinkDependencies)
+  }
+
+  /**
+   * Immutably remove a sink from this graph.
+   *
+   * @param sink The id of the sink to remove
+   * @return The new graph
+   */
+  def removeSink(sink: SinkId): Graph = {
+    require(sinkDependencies.contains(sink), "Sink being removed must exist")
+
+    val newSinkDependencies = sinkDependencies - sink
+    copy(sinkDependencies = newSinkDependencies)
+  }
+
+  /**
+   * Immutably remove a source from this graph.
+   * Note: This may leave invalid dangling dependencies on the deleted source,
+   * that must be manually taken care of.
+   *
+   * @param source The id of the source to remove
+   * @return The new graph
+   */
+  def removeSource(source: SourceId): Graph = {
```

Comment:
That's what I was thinking - though an argument for API symmetry could be made. I have a preference for smaller API surface area over symmetry, but your call if you think programming against `removeNode` is weird?

On another note - I haven't been looking hard for consistency of usage of `Node` vs. `Vertex`, etc. (though I think the only other place `vertex` is used is in the comments and I guess it should be updated?) - In general maybe take a pass looking for common sources of confusion and make sure that we're consistent with language here.